### DISCUSSION TITLE
Get `metadata` in `InvoicePaymentSettled` webhook?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [btcpayserver/btcpayserver#5633] [Technical Support] [answered]
I have 2 API routes:

1. `/api/btcpayserver/checkout` -> this one calls `${BTCPAY_URL}/stores/${BTCPAY_STORE_ID}/invoices` with `metadata` object
2. `/api/btcpayserver/webhook` -> i want to access `metadata` in this route

i can't find a way to do that. stripe works just like this.

i create a user in webhook when `InvoicePaymentSettled` event is called. that's why i need it. i thought about creating a user in `checkout` with `unpaid` status & turn it to `paid` in `webhook` but even then i need an email or some identifier.

i saw the api docs & it does show `metadata` in `InvoicePaymentSettled` -> https://docs.btcpayserver.org/API/Greenfield/v1/#operation/Webhook_InvoiceSettled

but i can't find it in all webhook events.

related issue that looks fixed but doesn't work for me -> https://github.com/btcpayserver/btcpayserver/issues/4223 (you can check it yourself as i've made a minimal reproduction -> https://github.com/deadcoder0904/btcpayserver-nextjs)

**Comment by ndeet** [comment]
Hi!

- which BTCPay Server version do you run?
- are you sure your metadata is set correctly and the invoice has it?

**Reply by deadcoder0904** [reply]
1. the latest one on clovyr as i created an account there <3 days ago. its written `© BTCPay Server v1.6.12.0` at the bottom
2. yes, you can see my minimal code here & detailed instructions which you already know but lmk if you don't work with `node.js` & aren't familiar with `pnpm` so i'll add instructions -> https://github.com/deadcoder0904/btcpayserver-nextjs (see the `api` folder. see `console.log({ event })` [here](https://github.com/deadcoder0904/btcpayserver-nextjs/blob/659e69fa22fd0238a045e8f5b2266b9a5a43dc48/src/app/api/btcpayserver/webhook/route.ts#L43) in `/api/btcpayserver/webhook/route.ts` that should print `metadata` which i've set [here](https://github.com/deadcoder0904/btcpayserver-nextjs/blob/659e69fa22fd0238a045e8f5b2266b9a5a43dc48/src/app/api/btcpayserver/checkout/route.ts#L52) in `/api/btcpayserver/checkout/route.ts` ... bdw the `metadata` is printed in the `/api/btcpayserver/checkout/route.ts` perfectly where i `console.log({ session})` [here](https://github.com/deadcoder0904/btcpayserver-nextjs/blob/659e69fa22fd0238a045e8f5b2266b9a5a43dc48/src/app/api/btcpayserver/checkout/route.ts#L59))

**Reply by deadcoder0904** [reply, marked as answer]
alright so i got it working by calling `${BTCPAY_URL}/stores/${BTCPAY_STORE_ID}/invoices/${BTCPAY_INVOICE_ID}` again which i think shouldn't be needed so this might be a bug. i'll open an issue for now.

```tsx
const response = await fetch(
			`${BTCPAY_URL}/stores/${BTCPAY_STORE_ID}/invoices/${BTCPAY_INVOICE_ID}`,
			{
				method: "GET",
				headers: {
					"Content-Type": "application/json",
					Authorization: `token ${BTCPAY_API_KEY}`,
				},
			}
		)

		const { metadata } = await response.json()
```

**Reply by ndeet** [reply]
Ok thanks for debugging, yes that invoice endpoint is always a solution but if the metadata should be present in the webhook payload, according to api docs, then it is a bug you are right. Will double check and confirm in your issue if I can reproduce with v1.12.3.

**Reply by ndeet** [reply]
I can't reproduce your problem tested with BTCPay v1.12.0 and v1.12.3, the metadata is present for me in all events, e.g. InvoicePaymentSettled event in the image
![image](https://github.com/btcpayserver/btcpayserver/assets/1136761/616d38e6-6c7f-4fff-9ecd-aa97cd428aba)

Please check the response again if you maybe access it wrong, or can you dump a response payload here? 

btw I would not use that event but `InvoiceSettled` as an invoice can have multiple payments and each payment will have an InvoicePaymentSettled event.

**Reply by deadcoder0904** [reply]
oh yes, @ndeet i already learnt that in https://github.com/btcpayserver/btcpayserver/issues/5637

thanks for `InvoiceSettled`... i'll use that from now on.

is there anything for failure event? or is it just expired?

i'd like to do something similar for failure like stripe has `"checkout.session.async_payment_failed"`. what's the btcpay equivalent?

**Reply by ndeet** [reply]
When you look at the available webhooks you can see that we have `InvoiceExpired` and `InvoiceInvalid` endpoints that can get triggered on expired or failed payments.
